onnxruntime/core/session/custom_ops.cc

```diff
@@ -402,12 +411,14 @@ common::Status CreateCustomRegistry(gsl::span<OrtCustomOpDomain* const> op_domai
 
         const auto type = op->GetOutputType(op, i);
         if (ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED == type) {  // Dynamic typed output
-          ORT_ENFORCE(type_id_counter == 1,
-                      "There must be one (and only one) dynamic typed input to the custom op. "
-                      "Its type info at runtime will be used to infer the type info of this dynamic typed output "
-                      "which is required for the success of the model loading step. "
-                      "More than one dynamic typed inputs are currently not supported as differing types at runtime means the output type "
-                      "cannot be inferred without which model loading cannot proceed.");
+          if (op->GetOutputCharacteristic(op, i) == OrtCustomOpInputOutputCharacteristic::INPUT_OUTPUT_REQUIRED) {
```

Comment:
for inputs with INPUT_OUTPUT_OPTIONAL characteristic, will their type info be needed to infer type info of dynamic typed output?

Reply:
According to the code in custom_ops.cc, regardless of which characteristic of the inputs, their type info will be captured. 
Whether optional input's type info is needed for inferring output's type later, the node schema should already have this information.